assign_rights/tests.py

```diff
@@ -95,21 +109,51 @@ def test_grouping_form(self):
 class TestRightsAssembler(TestCase):
     def setUp(self):
         add_rights_shells()
+        add_rights_acts()
         self.assembler = RightsAssembler()
+        self.rights_ids = [obj.pk for obj in RightsShell.objects.all()]
 
     def test_retrieve_rights(self):
         """Tests the retrieve_rights method.
 
         Asserts the method returns a list or DoesNotExist exception.
         """
-        rights_ids = [obj.pk for obj in RightsShell.objects.all()]
-        assembled = self.assembler.retrieve_rights(rights_ids)
+        assembled = self.assembler.retrieve_rights(self.rights_ids)
         self.assertTrue(isinstance(assembled, list))
-        self.assertEqual(len(rights_ids), len(assembled))
+        self.assertEqual(len(self.rights_ids), len(assembled))
 
-        rights_ids.append(len(rights_ids) + 1)
+        deleted = random.choice(RightsShell.objects.all())
+        deleted.delete()
+        self.rights_ids.append(deleted.pk)
         with self.assertRaises(RightsShell.DoesNotExist):
-            assembled = self.assembler.retrieve_rights(rights_ids)
+            assembled = self.assembler.retrieve_rights(self.rights_ids)
+
+    def check_end_dates(self, assembled, grouping_end_date):
```

Comment:
I've been thinking about this and I think that instead of randomly setting attributes on the object, you may want to explicitly set those and then check for expected outputs. So, for example, set `end_date_open` to True and then assert that `calculate_dates` returns `None`.

That way you can be sure that you're consistently testing all the branches of the logic, and it's also a little clearer for someone coming to it without prior knowledge.

Reply:
Ok, I've pushed up my latest changes. I'm not super happy with the tests right now, but I'm a little confused about how we'd go about doing the above while still keeping the random generation. With start dates added, there are somewhere like 6 different start date and end date permutations for each RightsShells and RightsGranted.

Reply:
Why don't you get the rest of this settled and then I can work with you on tests.